README.md
# NAME

Pod::Weaver::Section::Badges - Add (or append) a section with badges

<div>
    <p>
    <img src="https://img.shields.io/badge/perl-5.10+-blue.svg" alt="Requires Perl 5.10+" />
    <a href="https://travis-ci.org/Csson/p5-Pod-Weaver-Section-Badges"><img src="https://api.travis-ci.org/Csson/p5-Pod-Weaver-Section-Badges.svg?branch=master" alt="Travis status" /></a>
    <a href="http://cpants.cpanauthors.org/dist/Pod-Weaver-Section-Badges-0.0402"><img src="https://badgedepot.code301.com/badge/kwalitee/Pod-Weaver-Section-Badges/0.0402" alt="Distribution kwalitee" /></a>
    <a href="http://matrix.cpantesters.org/?dist=Pod-Weaver-Section-Badges%200.0402"><img src="https://badgedepot.code301.com/badge/cpantesters/Pod-Weaver-Section-Badges/0.0402" alt="CPAN Testers result" /></a>
    <img src="https://img.shields.io/badge/coverage-81.0%-orange.svg" alt="coverage 81.0%" />
    </p>
</div>

# VERSION

Version 0.0402, released 2016-02-20.

# SYNOPSIS

    ; in weaver.ini
    [Badges]
    section = BUILD STATUS
    formats = html
    badge = Travis
    badge = Gratipay
    -travis_user = MyGithubUser
    -travis_repo = the_repository
    -travis_branch = master
    -gratipay_user = ExampleName

# DESCRIPTION

This inserts a section with status badges. The configuration in the synopsis would produce something similar to this:

    =head1 BUILD STATUS

    =begin HTML

    <p>
        <a href="https://travis-ci.org/MyGithubUser/the_repository"><img src="https://travis-ci.org/MyGithubUser/the_repository.svg?branch=master" /></a>
        <img src="https://img.shields.io/gratipay/ExampleName.svg" />
    </p>

    =end HTML

This module uses badges in the `Badge::Depot::Plugin` namespace. See [Task::Badge::Depot](https://metacpan.org/pod/Task::Badge::Depot) for a list of available badges.
The synopsis uses the [Badge::Depot::Plugin::Travis](https://metacpan.org/pod/Badge::Depot::Plugin::Travis) and [Badge::Depot::Plugin::Gratipay](https://metacpan.org/pod/Badge::Depot::Plugin::Gratipay) badges.

Attributes starting with a dash (such as, in the synopsis, `-travis_user` or `-gratipay_user`) are given to each badge's constructor.

## Badge rendering

As a comparison with using badges and [Badge::Depot](https://metacpan.org/pod/Badge::Depot) directly, this is what `Pod::Weaver::Section::Badges` does.

First, with this part of the synopsis:

    [Badges]
    badge = Gratipay
    -gratipay_user = ExampleName

`badge = Gratipay` means that [Badge::Depot::Plugin::Gratipay](https://metacpan.org/pod/Badge::Depot::Plugin::Gratipay) is automatically `used`.

Secondly, `-gratipay_user = Example` means that this attribute is for the `Gratipay` badge, so the prefix (`-gratipay_`) is stripped and the attribute is given in the constructor:

    my $gratipay_badge = Badge::Depot::Plugin::Gratipay->new(user => 'ExampleName');

And then the given `formats` is used to render the pod:

    my $rendered_badge = $gratipay_badge->to_html;

Which is then injected into the chosen `section`.

# ATTRIBUTES

## formats

<table cellpadding="0" cellspacing="0">
<tr>
    <td style="padding-right: 6px; padding-left: 6px; border-right: 1px solid #b8b8b8; white-space: nowrap;"><a href="https://metacpan.org/pod/Types::Standard#ArrayRef">ArrayRef</a> [ <a href="https://metacpan.org/pod/Types::Standard#Enum">Enum</a> [ "<a href="https://metacpan.org/pod/Types::Standard#html">html</a>","<a href="https://metacpan.org/pod/Types::Standard#markdown">markdown</a>" ] ]</td>
    <td style="padding-right: 6px; padding-left: 6px; border-right: 1px solid #b8b8b8; white-space: nowrap;">required</td>
    <td style="padding-left: 6px; padding-right: 6px; white-space: nowrap;">read-only</td>
</tr>
</table>

<p>The formats to render the badges for. Comma separated list, not multiple rows.</p>

## badge

<table cellpadding="0" cellspacing="0">
<tr>
    <td style="padding-right: 6px; padding-left: 6px; border-right: 1px solid #b8b8b8; white-space: nowrap;"><a href="https://metacpan.org/pod/Types::Standard#ArrayRef">ArrayRef</a> [ <a href="https://metacpan.org/pod/Types::Standard#Str">Str</a> ]</td>
    <td style="padding-right: 6px; padding-left: 6px; border-right: 1px solid #b8b8b8; white-space: nowrap;">optional, default: <code>[]</code></td>
    <td style="padding-left: 6px; padding-right: 6px; white-space: nowrap;">read-only</td>
</tr>
</table>

<p>The name of the wanted badge, lowercased. Repeat for multiple badges. The name is everything after 'Badge::Depot::Plugin::'.</p>

## main\_module\_only

<table cellpadding="0" cellspacing="0">
<tr>
    <td style="padding-right: 6px; padding-left: 6px; border-right: 1px solid #b8b8b8; white-space: nowrap;"><a href="https://metacpan.org/pod/Types::Standard#Bool">Bool</a></td>
    <td style="padding-right: 6px; padding-left: 6px; border-right: 1px solid #b8b8b8; white-space: nowrap;">optional, default: <code>1</code></td>
    <td style="padding-left: 6px; padding-right: 6px; white-space: nowrap;">read-only</td>
</tr>
</table>

<p>If true, the badges will only be inserted in the main module (as defined by Dist::Zilla). If false, they will be included in all modules.</p>

## section

<table cellpadding="0" cellspacing="0">
<tr>
    <td style="padding-right: 6px; padding-left: 6px; border-right: 1px solid #b8b8b8; white-space: nowrap;"><a href="https://metacpan.org/pod/Types::Standard#Str">Str</a></td>
    <td style="padding-right: 6px; padding-left: 6px; border-right: 1px solid #b8b8b8; white-space: nowrap;">optional, default: <code>NAME</code></td>
    <td style="padding-left: 6px; padding-right: 6px; white-space: nowrap;">read-only</td>
</tr>
</table>

<p>The section of pod to add the badges to, identified by its heading. The section will be created if it doesn't already exist.</p>

## skip\_markdown\_if\_html

<table cellpadding="0" cellspacing="0">
<tr>
    <td style="padding-right: 6px; padding-left: 6px; border-right: 1px solid #b8b8b8; white-space: nowrap;"><a href="https://metacpan.org/pod/Types::Standard#Bool">Bool</a></td>
    <td style="padding-right: 6px; padding-left: 6px; border-right: 1px solid #b8b8b8; white-space: nowrap;">optional, default: <code>1</code></td>
    <td style="padding-left: 6px; padding-right: 6px; white-space: nowrap;">read-only</td>
</tr>
</table>

<p>Some markdown renderers also renders '=begin html' blocks, which makes it unnecessary to set both html and markdown as output formats. Set this to a false value to produce both blocks.</p>

# SEE ALSO

- [Task::Badge::Depot](https://metacpan.org/pod/Task::Badge::Depot)
- [Badge::Depot](https://metacpan.org/pod/Badge::Depot)

# BADGES

![](https://img.shields.io/badge/perl-5.14+-brightgreen.svg)
    [![](https://travis-ci.org/Csson/p5-Pod-Weaver-Section-Badges.svg?branch=master)](https://travis-ci.org/Csson/p5-Pod-Weaver-Section-Badges)

# SOURCE

[https://github.com/Csson/p5-Pod-Weaver-Section-Badges](https://github.com/Csson/p5-Pod-Weaver-Section-Badges)

# HOMEPAGE

[https://metacpan.org/release/Pod-Weaver-Section-Badges](https://metacpan.org/release/Pod-Weaver-Section-Badges)

# AUTHOR

Erik Carlsson <info@code301.com>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2016 by Erik Carlsson.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for Pod-Weaver-Section-Badges

{{$NEXT}}
 [API Changes]

 [Bug Fixes]

 [Documentation]

 [Enhancements]

 [New Features]

 [Other Changes]

 [Requirements]

 [Stats & Dependency Changes]

 [Testing]

0.0402 2016-02-20T20:29:12Z UTC
 [Enhancements]
 - Pass on the Dist::Zilla object (if available) to the badges.
 - Don't create any output if the badge plugin doesn't return anything.
   Previously created a line break.

 [Stats & Dependency Changes]
 - 6 files changed, 75 insertions(+), 16 deletions(-)
 - (run req) + List::Util 1.42
 - (run req) + PerlX::Maybe 1.001
 - (run req) ~ Moose 2.1400 --> 2.1405
 - (run req) ~ MooseX::AttributeDocumented (any) --> 0.1003
 - (run req) ~ Types::Standard 1.000000 --> 1.000005
 - (run req) ~ namespace::autoclean 0.18 --> 0.22
 - (run req) - List::AllUtils
 - (test req) ~ Badge::Depot (any) --> 0.0104
 - (test req) ~ Path::Tiny (any) --> 0.072
 - (dev req) + Badge::Depot 0.0104
 - (dev req) + Badge::Depot::Plugin::Coverage (any)
 - (dev req) + Badge::Depot::Plugin::Cpantesters (any)
 - (dev req) + Badge::Depot::Plugin::Kwalitee (any)
 - (dev req) + Badge::Depot::Plugin::Perl (any)
 - (dev req) + Badge::Depot::Plugin::Travis (any)
 - (dev req) + Dist::Zilla 5.000
 - (dev req) + Dist::Zilla::Plugin::Git 2.036
 - (dev req) + Pod::Weaver::Section::Badges 0.0401
 - (dev req) + Pod::Weaver::Section::GenerateSection 1.01
 - (dev req) + Test::EOF (any)
 - (dev req) + Test::Version 1
 - (dev req) ~ Dist::Zilla::Plugin::ChangeStats::Dependencies::Git (any)
   --> 0.0100
 - (dev req) ~ Dist::Zilla::Plugin::CheckChangesHasContent (any) --> 0.008
 - (dev req) ~ Dist::Zilla::Plugin::Clean (any) --> 0.07
 - (dev req) ~ Dist::Zilla::Plugin::CopyFilesFromBuild (any) --> 0.151680
 - (dev req) ~ Dist::Zilla::Plugin::Git::Contributors (any) --> 0.017
 - (dev req) ~ Dist::Zilla::Plugin::GithubMeta (any) --> 0.54
 - (dev req) ~ Dist::Zilla::Plugin::InstallRelease (any) --> 0.008
 - (dev req) ~ Dist::Zilla::Plugin::MetaProvides::Package (any) -->
   2.003001
 - (dev req) ~ Dist::Zilla::Plugin::NextRelease::Grouped (any) --> 0.0200
 - (dev req) ~ Dist::Zilla::Plugin::PodWeaver (any) --> 4.006
 - (dev req) ~ Dist::Zilla::Plugin::ReadmeAnyFromPod (any) --> 0.150250
 - (dev req) ~ Dist::Zilla::Plugin::RunExtraTests (any) --> 0.028
 - (dev req) ~ Dist::Zilla::Plugin::Test::Compile (any) --> 2.054
 - (dev req) ~ Dist::Zilla::Plugin::Test::EOF (any) --> 0.0501
 - (dev req) ~ Dist::Zilla::Plugin::Test::EOL (any) --> 0.18
 - (dev req) ~ Dist::Zilla::Plugin::Test::Kwalitee::Extra (any) --> v0.2.1
 - (dev req) ~ Dist::Zilla::Plugin::Test::NoTabs (any) --> 0.15
 - (dev req) ~ Dist::Zilla::Plugin::Test::ReportPrereqs (any) --> 0.021
 - (dev req) ~ Dist::Zilla::Plugin::Test::Version (any) --> 1.05
 - (dev req) ~ Dist::Zilla::Plugin::TravisYML (any) --> 1.15
 - (dev req) ~ Test::NoTabs (any) --> 1.4
 - (dev req) ~ Test::Warnings (any) --> 0.026
 - (dev req) - Dist::Zilla::Plugin::MetaProvides::Class
 - (dev req) - Dist::Zilla::Plugin::PodnameFromClassname
 - (dev req) - Dist::Zilla::Plugin::Prereqs::Plugins
 - (dev req) - Pod::Weaver::Section::Homepage::DefaultCPAN
 - (dev req) - Pod::Weaver::Section::Source::DefaultGitHub
 - (dev sug) ~ Dist::Iller 0.1406 --> 0.1407
 - (dev sug) ~ Dist::Iller::Config::Author::CSSON 0.0307 --> 0.0309

0.0401 2016-02-17T22:42:47Z UTC
 [Enhancements]
 - New attribute 'skip_markdown_if_html' doesn't produce '=begin markdown'
   blocks if both html and markdown are set as formats. Some (all?)
   markdown renderers also render html blocks. This avoids duplicating the
   badges.
 - Badges are now rendered one-per-line. This ensures that lines aren't
   overly long when using many badges.

 [Stats & Dependency Changes]
 - 6 files changed, 44 insertions(+), 24 deletions(-)
 - (run req) ~ perl 5.010001 --> 5.010000
 - (dev req) + Dist::Zilla::Plugin::Authority 1.009
 - (dev req) + Dist::Zilla::Plugin::BumpVersionAfterRelease::Transitional
   0.007
 - (dev req) + Dist::Zilla::Plugin::ChangeStats::Dependencies::Git (any)
 - (dev req) + Dist::Zilla::Plugin::Git::CheckFor::CorrectBranch 0.013
 - (dev req) + Dist::Zilla::Plugin::Git::Commit (any)
 - (dev req) + Dist::Zilla::Plugin::NextRelease::Grouped (any)
 - (dev req) + Dist::Zilla::Plugin::RewriteVersion::Transitional 0.007
 - (dev req) ~ Pod::Elemental::Transformer::Splint (any) --> 0.1201
 - (dev req) ~ Test::More (any) --> 0.88
 - (dev req) - Dist::Iller::Config::Author::CSSON
 - (dev req) - Dist::Zilla::Plugin::NextRelease
 - (dev req) - Dist::Zilla::Plugin::NextVersion::Semantic
 - (dev req) - Dist::Zilla::Plugin::OurPkgVersion
 - (dev req) - Dist::Zilla::Plugin::PreviousVersion::Changelog
 - (dev req) - Dist::Zilla::Plugin::ReversionOnRelease
 - (dev sug) + Dist::Iller 0.1406
 - (dev sug) + Dist::Iller::Config::Author::CSSON 0.0307

0.0400 2016-01-27T22:16:34Z UTC
 [Enhancements]
 - Can now be used without Dist::Zilla (#3, thanks MGRIMES).

0.0300 2016-01-26T18:25:47Z UTC
 [Other Changes]
 - Rebuilt with current Dist::Iller::Config::Author::CSSON.

 [Requirements]
 - Remove Moops dependency (should now work on 5.10+).
 - New (explicit) prereqs: namespace::autoclean, Types::Standard.

 [Testing]
 - Fix year dependent test (#2, thanks SREZIC).

0.0203 2015-04-11T23:58:00Z CEST
 [Documentation]
 - Removed documentation for the internal 'badge_args' attribute, and
   instead:
 - Added a section of documentation describing a bit more how badges are
   rendered given the synopsis (and compared to using Badge::Depot
   directly).

0.0202 2015-03-23T18:01:15Z CET
 [Enhancements]
 - Renamed the plugin searcher class so that it is under the
   Pod::Weaver::Section::Badges namespace.

 [Requirements]
 - Runtime: Listed List::AllUtils
 - Test: Listed Syntax::Feature::Qi

0.0201 2015-03-21T16:56:08Z CET
 [Documentation]
 - Corrected abstract.

0.0200 2015-03-21T16:49:20Z CET
 [API Changes]
 - This version has a completely new (and better) way to choose badges:

t/corpus/01/lib/TesterFor/Badges.pm
use 5.14.0;
use strict;
use warnings;
use Moops;

# PODNAME: TesterFor::Badges
# ABSTRACT: A tester

class TesterFor::Badges using Moose with Pod::Weaver::Section::Badges::Utils {
    has badge_args => (
        is => 'ro',
        isa => HashRef[Str],
        default => sub { {} },
        traits => ['Hash'],
        handles => {
            badge_args_kv => 'kv',
        },
    );
}

1;

__END__

=pod

=head1 ATTRIBUTES

Some attributes.

=cut


t/corpus/01/lib/Badge/Depot/Plugin/Anothertestplugin.pm
use 5.10.1;
use strict;
use warnings;

package Badge::Depot::Plugin::Anothertestplugin;

# VERSION

use Moose;
use Types::Standard qw/Str/;
with 'Badge::Depot';

has user => (
    is => 'ro',
    isa => Str,
);
has repo => (
    is => 'ro',
    isa => Str,
);
    
sub BUILD {
    my $self = shift;
    $self->link_url(sprintf 'https://example.org/%s/%s' => $self->user, $self->repo);
    $self->image_url(sprintf 'https://example.org/%s/%s.svg' => $self->user, $self->repo);
}

1;


t/corpus/01/lib/Badge/Depot/Plugin/Thisisatestplugin.pm
use 5.10.1;
use strict;
use warnings;

package Badge::Depot::Plugin::Thisisatestplugin;

# VERSION

use Moose;
use Types::Standard qw/Str/;
with 'Badge::Depot';

has user => (
    is => 'ro',
    isa => Str,
);
has repo => (
    is => 'ro',
    isa => Str,
);

sub BUILD {
    my $self = shift;
    $self->link_url(sprintf 'https://example.com/%s/%s' => $self->user, $self->repo);
    $self->image_url(sprintf 'https://example.com/%s/%s.svg' => $self->user, $self->repo);
}

1;


lib/Pod/Weaver/Section/Badges.pm
use 5.10.0;
use strict;
use warnings;

package Pod::Weaver::Section::Badges;

# ABSTRACT: Add (or append) a section with badges
# AUTHORITY
our $VERSION = '0.0403';

use Moose;
use MooseX::AttributeDocumented;
use PerlX::Maybe qw/provided/;
use Types::Standard -types;
use namespace::autoclean;
use Pod::Weaver::Section::Badges::PluginSearcher;
with qw/
    Pod::Weaver::Role::Section
    Pod::Weaver::Role::AddTextToSection
    Pod::Weaver::Section::Badges::Utils
/;

sub mvp_multivalue_args { qw/badge/ }

has '+weaver' => (
    is => 'ro',
    documentation_order => 0,
    traits => ['Documented'],
);
has '+logger' => (
    is => 'ro',
    traits => ['Documented'],
    documentation_order => 0,
);
has '+plugin_name' => (
    is => 'ro',
    documentation_order => 0,
    traits => ['Documented'],
);

has badge => (
    is => 'ro',
    isa => ArrayRef[Str],
    traits => ['Array'],
    default => sub { [] },
    handles => {
        all_badges => 'elements',
        find_badge => 'first',
        count_badges => 'count',
    },
    documentation => q{The name of the wanted badge, lowercased. Repeat for multiple badges. The name is everything after 'Badge::Depot::Plugin::'.},
    documentation_default => '[]',
);
has section => (
    is => 'ro',
    isa => Str,
    default => 'NAME',
    documentation => q{The section of pod to add the badges to, identified by its heading. The section will be created if it doesn't already exist.},
);
has formats => (
    is => 'ro',
    isa => ArrayRef[Enum[qw/html markdown/]],
    traits => ['Array'],
    required => 1,
    handles => {
        all_formats => 'elements',
        find_format => 'first',
        has_formats => 'count',
    },
    documentation => q{The formats to render the badges for. Comma separated list, not multiple rows.},
    documentation_default => '[]',
);
has skip_markdown_if_html => (
    is => 'ro',
    isa => Bool,
    default => 1,
    documentation => q{Some markdown renderers also renders '=begin html' blocks, which makes it unnecessary to set both html and markdown as output formats. Set this to a false value to produce both blocks.},
);

has badge_args => (
    is => 'ro',
    isa => HashRef[Str],
    default => sub { {} },
    traits => ['Hash'],
    handles => {
        badge_args_kv => 'kv',
    },
    documentation_order => 0,
);
has plugin_searcher => (
    is => 'ro',
    isa => Any,
    init_arg => undef,
    default => sub { Pod::Weaver::Section::Badges::PluginSearcher->new },
    documentation_order => 0,
);
has main_module_only => (
    is => 'ro',
    isa => Bool,
    default => 1,
    documentation => 'If true, the badges will only be inserted in the main module (as defined by Dist::Zilla). If false, they will be included in all modules.',
);


around BUILDARGS => sub {
    my $next = shift;
    my $class = shift;
    my $args = shift;
    $args->{'formats'} = exists $args->{'formats'} ? [ split /, ?/ => $args->{'formats'} ] : [];
    $args->{'badge_args'} = { map { $_ => delete $args->{ $_  } } grep { /^-/ } keys %$args };

    $class->$next($args);
};

sub main_module {
    my $self = shift;
    my $input = shift;

    # Use the main_module as defined by Dist::Zilla if it is available
    return $input->{'zilla'}->main_module->name if exists $input->{'zilla'};

    # Taken from Dist::Zilla
    # If your distribution is Foo-Bar, and lib/Foo/Bar.pm exists, that's the main_module.
    (my $guess = 'lib/' . $input->{'meta'}{'name'} . '.pm') =~ s{-}{/}g;
    return -e $guess ? $guess : undef;
}

sub weave_section {
    my $self = shift;
    my $document = shift;
    my $input = shift;

    return if $self->main_module_only && $input->{'filename'} ne $self->main_module($input);

    my $badges_args = {
        provided ref $input eq 'HASH' && exists $input->{'zilla'}, zilla => $input->{'zilla'},
    };
    my $badge_objects = $self->create_badges($badges_args);
    return if !scalar @$badge_objects;

    if(!$self->has_formats) {
        $self->log(['!!! No formats defined in weaver.ini, no badges to create.']);
    }

    my $formats = [
        {
            name => 'html',
            before => '<p>',
            after => '</p>',
        },
        {
            name => 'markdown',
            before => undef,
            after => undef,
        }
    ];

    my @output = ();
    FORMAT:
    foreach my $format (@$formats) {
        # Optionally skip markdown if we also prints to html
        next FORMAT if $format->{'name'} eq 'markdown' && $self->find_format(sub { $_ eq 'html'}) && $self->skip_markdown_if_html;
        push @output => @{ $self->render_badges($format, $badge_objects) };
    }

    if(scalar @output) {
        my $output = join "\n" => '', @output, '';

        $self->add_text_to_section($document, $output, $self->section);
    }
}

__PACKAGE__->meta->make_immutable;

1;

__END__

=pod

:splint classname Pod::Weaver::Section::Badges

=head1 SYNOPSIS

    ; in weaver.ini
    [Badges]
    section = BUILD STATUS
    formats = html
    badge = Travis
    badge = Gratipay
    -travis_user = MyGithubUser
    -travis_repo = the_repository
    -travis_branch = master
    -gratipay_user = ExampleName
    
=head1 DESCRIPTION

This inserts a section with status badges. The configuration in the synopsis would produce something similar to this:

    =head1 BUILD STATUS

    =begin HTML

    <p>
        <a href="https://travis-ci.org/MyGithubUser/the_repository"><img src="https://travis-ci.org/MyGithubUser/the_repository.svg?branch=master" /></a>
        <img src="https://img.shields.io/gratipay/ExampleName.svg" />
    </p>

    =end HTML

This module uses badges in the C<Badge::Depot::Plugin> namespace. See L<Task::Badge::Depot> for a list of available badges.
The synopsis uses the L<Badge::Depot::Plugin::Travis> and L<Badge::Depot::Plugin::Gratipay> badges.

Attributes starting with a dash (such as, in the synopsis, C<-travis_user> or C<-gratipay_user>) are given to each badge's constructor.

=head2 Badge rendering

As a comparison with using badges and L<Badge::Depot> directly, this is what C<Pod::Weaver::Section::Badges> does.

First, with this part of the synopsis:

    [Badges]
    badge = Gratipay
    -gratipay_user = ExampleName

C<badge = Gratipay> means that L<Badge::Depot::Plugin::Gratipay> is automatically C<used>.

Secondly, C<-gratipay_user = Example> means that this attribute is for the C<Gratipay> badge, so the prefix (C<-gratipay_>) is stripped and the attribute is given in the constructor:

    my $gratipay_badge = Badge::Depot::Plugin::Gratipay->new(user => 'ExampleName');

And then the given C<formats> is used to render the pod:

    my $rendered_badge = $gratipay_badge->to_html;

Which is then injected into the chosen C<section>.

=head1 ATTRIBUTES

:splint attributes

=head1 SEE ALSO

=for :list
* L<Task::Badge::Depot>
* L<Badge::Depot>

=head1 BADGES

=begin HTML

<p>
    <img src="https://img.shields.io/badge/perl-5.10.1+-brightgreen.svg" />
    <a href="https://travis-ci.org/Csson/p5-Pod-Weaver-Section-Badges"><img src="https://travis-ci.org/Csson/p5-Pod-Weaver-Section-Badges.svg?branch=master" /></a>
</p>

=end HTML

=begin markdown

    ![](https://img.shields.io/badge/perl-5.14+-brightgreen.svg)
    [![](https://travis-ci.org/Csson/p5-Pod-Weaver-Section-Badges.svg?branch=master)](https://travis-ci.org/Csson/p5-Pod-Weaver-Section-Badges)

=end markdown

=cut


lib/Pod/Weaver/Section/Badges/Utils.pm
use 5.10.0;
use strict;
use warnings;

package Pod::Weaver::Section::Badges::Utils;

# ABSTRACT: Some helpers
# AUTHORITY
our $VERSION = '0.0403';

use Moose::Role;
use List::Util 'first';

sub get_params_for {
    my $self = shift;
    my $badgename = shift; # Str
    my $params = {};

    foreach my $pair ($self->badge_args_kv) {
        my($key, $value) = @$pair;

        # -pluginname_* -> *
        my $key_for_plugin = $key;
        $key_for_plugin =~ s{^-${badgename}_}{};

        $params->{ $key_for_plugin } = $value if substr($key, 0, 2 + length $badgename) eq sprintf '-%s_' => $badgename;
    }
    return %{ $params };
}

sub badge_to_class {
    my $self = shift;
    my $badge_name = shift; # Str
    return sprintf 'Badge::Depot::Plugin::%s', ucfirst $badge_name;
}

sub create_badges {
    my $self = shift;
    my $badges_args = shift || {};

    my @all_plugins = $self->plugin_searcher->plugins;
    my @badge_objects = ();

    BADGE:
    foreach my $badge ($self->all_badges) {
        my $wanted_plugin_class = $self->badge_to_class($badge);
        my $plugin_class = first { $_ eq $wanted_plugin_class } @all_plugins;

        next BADGE if !defined $plugin_class;
        my $plugin = $plugin_class->new($self->get_params_for(lc $badge), %{ $badges_args });

        if(!$plugin->DOES('Badge::Depot')) {
            warn sprintf '! %s does not consume the Badge::Depot role', $plugin_class;
            next BADGE;
        }
        push @badge_objects => $plugin;
    }
    return \@badge_objects; # ArrayRef[ ConsumerOf['Badge::Depot'] ]
}

sub render_badges {
    my $self = shift;
    my $format = shift;  # Dict[ name => Str, before => Maybe[Str], after => Maybe[Str] ]
    my $badges = shift;  # ArrayRef[ ConsumerOf['Badge::Depot'] ]

    my $pod_command_begin = sprintf '=begin %s', $format->{'name'};
    my $pod_command_end   = sprintf '=end %s', $format->{'name'};
    my $format_method = sprintf 'to_%s', $format->{'name'};

    my @badges_output = ();
    my @complete_output = ();

    if($self->find_format(sub { $_ eq $format->{'name'} })) {
        push @badges_output => grep { defined $_ && length $_ } $_->$format_method foreach (@$badges);

    }
    if(@badges_output) {
        push @complete_output => '', $pod_command_begin, '', ;
        push @complete_output => (($format->{'before'} || ''), join ("\n" => @badges_output), ($format->{'after'} || ''));
        push @complete_output => '', $pod_command_end, '';
    }
    return \@complete_output;
}


1;


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.043.
use strict;
use warnings;

use 5.010000;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Add (or append) a section with badges",
  "AUTHOR" => "Erik Carlsson <info\@code301.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Pod-Weaver-Section-Badges",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.010000",
  "NAME" => "Pod::Weaver::Section::Badges",
  "PREREQ_PM" => {
    "List::Util" => "1.42",
    "Module::Pluggable" => "4.7",
    "Moose" => "2.1405",
    "MooseX::AttributeDocumented" => "0.1003",
    "PerlX::Maybe" => "1.001",
    "Pod::Weaver::Role::AddTextToSection" => 0,
    "Pod::Weaver::Role::Section" => 0,
    "Types::Standard" => "1.000005",
    "namespace::autoclean" => "0.22"
  },
  "TEST_REQUIRES" => {
    "Badge::Depot" => "0.0104",
    "Dist::Zilla::Plugin::PodWeaver" => 0,
    "Dist::Zilla::Tester" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "File::pushd" => 0,
    "FindBin" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "PPI" => 0,
    "Path::Tiny" => "0.072",
    "Pod::Elemental" => 0,
    "Pod::Weaver" => 0,
    "Syntax::Feature::Qi" => "0.2004",
    "Test::DZil" => 0,
    "Test::Differences" => 0,
    "Test::More" => "0.96",
    "syntax" => 0
  },
  "VERSION" => "0.0403",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Badge::Depot" => "0.0104",
  "Dist::Zilla::Plugin::PodWeaver" => 0,
  "Dist::Zilla::Tester" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "File::pushd" => 0,
  "FindBin" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "List::Util" => "1.42",
  "Module::Pluggable" => "4.7",
  "Moose" => "2.1405",
  "MooseX::AttributeDocumented" => "0.1003",
  "PPI" => 0,
  "Path::Tiny" => "0.072",
  "PerlX::Maybe" => "1.001",
  "Pod::Elemental" => 0,
  "Pod::Weaver" => 0,
  "Pod::Weaver::Role::AddTextToSection" => 0,
  "Pod::Weaver::Role::Section" => 0,
  "Syntax::Feature::Qi" => "0.2004",
  "Test::DZil" => 0,
  "Test::Differences" => 0,
  "Test::More" => "0.96",
  "Types::Standard" => "1.000005",
  "namespace::autoclean" => "0.22",
  "syntax" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
